bsp/test/src/BspModulesTests.scala

```diff
@@ -0,0 +1,34 @@
+import mill.bsp.BSP
+import mill.util.ScriptTestSuite
+import utest._
+
+object BspModulesTests extends ScriptTestSuite(false) {
+  override def workspaceSlug: String = "bsp-modules"
+  override def scriptSourcePath: os.Path = os.pwd / "bsp" / "test" / "resources" / workspaceSlug
+
+  def tests: Tests = Tests {
+    test("BSP module with foreign modules") {
+      test("can be installed") {
+        val workspacePath = initWorkspace()
+        eval("mill.bsp.BSP/install") ==> true
+        os.exists(workspacePath / ".bsp" / s"${BSP.serverName}.json") ==> true
+      }
+      test("ModuleUtils resolves all referenced transitive modules") {
+        val workspacePath = initWorkspace()
+        eval("validate") ==> true
+        val file = workspacePath / "out" / "validate.dest" / "transitive-modules.json"
+        os.exists(file) ==> true
+        val readModules = os.read.lines(file).sorted
+        val expectedModules = Seq(
+          "", // the root module has no segemnts at all
+          "HelloBsp",
+          "HelloBsp.test",
+          "foreign-modules.proj1.proj1",
+          "foreign-modules.proj2.proj2",
+          // "proj3" // still not detected
```

Comment:
I have no idea how to detect this. We could reflect all targets of known modules and inspect their inputs, but although we have their segments, there is no way to find the owning modules.

Reply:
No idea either..

Reply:
The expected project name was incorrect.

```suggestion
          // "foreign-modules.proj3.proj3" // still not detected
```

Reply:
Maybe a module could register itself in its constructor in some sort of global `identifier => module` Map.
Every module when it first gets instantiated registers itself in the Mill "runtime".
Could this be a thing?
Probably it can be generalized to tasks as well.

Reply:
I'd rather avoid such global state. This will get in the way for any dynamically created module/target and also effectively impedes Mill usage as a library. It may even break our tests.

I think we should just find a way to add that missing information when processing the `import $file` lines. Then we can expose this info through a (synthetic) target in the main/base module or via API.